cloud/scope/vpc_cluster.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/IBM/platform-services-go-sdk/resourcecontrollerv2"
+
+	"k8s.io/klog/v2/textlogger"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	capiv1beta1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/patch"
+
+	infrav1beta2 "sigs.k8s.io/cluster-api-provider-ibmcloud/api/v1beta2"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/authenticator"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/cos"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcecontroller"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/resourcemanager"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/cloud/services/vpc"
+	"sigs.k8s.io/cluster-api-provider-ibmcloud/pkg/endpoints"
+)
+
+const (
+	// LOGDEBUGLEVEL indicates the debug level of the logs.
+	LOGDEBUGLEVEL = 5
+)
+
+// VPCClusterScopeParams defines the input parameters used to create a new VPCClusterScope.
+type VPCClusterScopeParams struct {
+	Client          client.Client
+	Cluster         *capiv1beta1.Cluster
+	IBMVPCCluster   *infrav1beta2.IBMVPCCluster
+	Logger          logr.Logger
+	ServiceEndpoint []endpoints.ServiceEndpoint
+
+	IBMVPCClient vpc.Vpc
+}
+
+// VPCClusterScope defines a scope defined around a VPC Cluster.
+type VPCClusterScope struct {
+	logr.Logger
+	Client      client.Client
+	patchHelper *patch.Helper
+
+	COSClient                cos.Cos
+	ResourceControllerClient resourcecontroller.ResourceController
+	ResourceManagerClient    resourcemanager.ResourceManager
+	VPCClient                vpc.Vpc
+
+	Cluster         *capiv1beta1.Cluster
+	IBMVPCCluster   *infrav1beta2.IBMVPCCluster
+	ServiceEndpoint []endpoints.ServiceEndpoint
+}
+
+// NewVPCClusterScope creates a new VPCClusterScope from the supplied parameters.
+func NewVPCClusterScope(params VPCClusterScopeParams) (*VPCClusterScope, error) {
+	if params.Client == nil {
+		err := errors.New("error failed to generate new scope from nil Client")
+		return nil, err
+	}
+	if params.Cluster == nil {
+		err := errors.New("error failed to generate new scope from nil Cluster")
+		return nil, err
+	}
+	if params.IBMVPCCluster == nil {
+		err := errors.New("error failed to generate new scope from nil IBMVPCCluster")
+		return nil, err
+	}
+	if params.Logger == (logr.Logger{}) {
+		params.Logger = textlogger.NewLogger(textlogger.NewConfig())
+	}
+
+	helper, err := patch.NewHelper(params.IBMVPCCluster, params.Client)
+	if err != nil {
+		return nil, fmt.Errorf("error failed to init patch helper: %w", err)
+	}
+
+	vpcEndpoint := endpoints.FetchVPCEndpoint(params.IBMVPCCluster.Spec.Region, params.ServiceEndpoint)
+	vpcClient, err := vpc.NewService(vpcEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("error failed to create IBM VPC client: %w", err)
+	}
+
+	if params.IBMVPCCluster.Spec.Network == nil || params.IBMVPCCluster.Spec.Region == "" {
+		return nil, fmt.Errorf("error failed to generate vpc client as NetworkSpec info is nil")
```

Comment:
if statement is checking for Region but it is not reported in the error, this may mislead the user..

Reply:
updated error info